### PR TITLE
Add `bind_async()` and `bind_tokio()`

### PR DESCRIPTION
**Pull request by simonbuchan**
Implementation of #16.

Still draft, as needs more complete docs on the new APIs.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,6 +8,10 @@ license = "MIT OR Apache-2.0"
 readme = "README.md"
 repository = "https://github.com/Srinivasa314/alcro"
 
+[[test]]
+name = "integration_test"
+required-features = ["tokio"]
```
Comment (Srinivasa314) [inline Cargo.toml:16]:
```suggestion
```
Tokio shouldnt be required for all tests

File: tests/integration_test.rs [hunk 2]
```diff
@@ -32,3 +32,54 @@ fn test_eval() {
     assert_eq!(ui.eval("throw 'ERROR'").unwrap_err(), "ERROR");
     assert!(ui.eval("dtyfhgxnt*").is_err());
 }
+
+#[tokio::test]
```
Comment (Srinivasa314) [inline tests/integration_test.rs]:
```suggestion
#[test]
```
No need of running this through tokio

File: tests/integration_test.rs [hunk 3]
```diff
@@ -32,3 +32,54 @@ fn test_eval() {
...
+async fn test_bind_async() {
```
Comment (Srinivasa314) [inline tests/integration_test.rs]:
```suggestion
fn test_bind_async() {
```

File: tests/integration_test.rs [hunk 4]
```diff
@@ -32,3 +32,54 @@ fn test_eval() {
...
+        .run()
+        .expect("Unable to launch");
+
+    ui.bind_async("bar", move |context| {
+        std::thread::Builder::new()
+            .name("test_bind_async binding".into())
+            .spawn(move || {
+                let result = format!("{}c", context.args()[0].as_str().expect("arg to be str"));
+                context.complete(Ok(result.into()))
+            });
```
Comment (Srinivasa314) [inline tests/integration_test.rs]:
```suggestion
            }).expect("Cannot spawn thread");
```
Do error checking

File: tests/integration_test.rs [hunk 5]
```diff
@@ -32,3 +32,54 @@ fn test_eval() {
...
+        .custom_args(&["--headless"])
+        .run()
+        .expect("Unable to launch");
+
+    ui.bind_tokio("bar", move |args| async move {
+        tokio::task::yield_now().await;
+        Ok(format!("{}c", args[0].as_str().expect("arg to be str")).into())
+    }).unwrap();
+
+    assert_eq!(ui.eval("foo('a')").unwrap(), "abcd");
```
Comment (Srinivasa314) [inline tests/integration_test.rs]:
```suggestion
    tokio::task::spawn_blocking(move || assert_eq!(ui.eval("foo('a')").unwrap(), "abcd")).await.unwrap();
```
Check result of foo('a') in non blocking manner

**Review by Srinivasa314 — CHANGES_REQUESTED** [review]
Also it would be good if you run rustfmt

**Comment by simonbuchan** [comment]
Thanks for the review, and sorry about the rustfmt! I thought my editor had it enabled by default already.

**Comment by Srinivasa314** [comment]
What do you think we can do with this PR?

**Comment by simonbuchan** [comment]
Oh dear, I certainly didn't follow up on this.

I feel this is *probably* usable - it would need proper documentation written for it, but I feel the bigger issue is if you're happy with the trade-offs and commitment to this specific API (eg., if the internals were async-ified later). Given that I think I can polish off the docs tomorrow and un-draft this?

**Comment by Srinivasa314** [comment]
> Given that I think I can polish off the docs tomorrow and un-draft this?

Yes

File: Cargo.toml (hunk 1 above)
Comment (simonbuchan) [reply Cargo.toml:16]: Think this is from an older version?

**Pull request merged**